Implement Pickpocket

ghoulslash opened the pull request:
##  Description
- Implement ABILITY_PICKPOCKET effect, which allows the target to steal the attacker's item if a contact move was used.
- occurs just before dancer in the move-end sequence
- Berry effects occur after pickpocket

## Additional Notes
- added `B_TRAINERS_STEAL_ITEMS` feature, where trainers can steal/swap items with the player. Non-berry items are restored afterwards. This makes trainers potentially more difficult/strategic, imo.


![pickpocket](https://user-images.githubusercontent.com/41651341/99036545-175bed80-253f-11eb-80d9-f198abd7d282.gif)

ghoulslash commented:
ready for review. Tested with double battles/trainer battles with no issues

AsparagusEduardo commented on src/battle_main.c:
```diff
@@ -4837,7 +4841,7 @@ static void HandleEndTurn_FinishBattle(void)
 {
     if (!gPaletteFade.active)
-    {
+    {        
```
```suggestion
    {
```

AsparagusEduardo commented on include/constants/battle_config.h:
```diff
@@ -133,6 +133,7 @@
 #define B_PARALYZE_ELECTRIC         GEN_6 // In Gen6+, Electric-type Pokémon can't be paralyzed.
 #define B_POWDER_GRASS              GEN_6 // In Gen6+, Grass-type Pokémon are immune to powder and spore moves.
 #define B_STEEL_RESISTANCES         GEN_6 // In Gen6+, Steel-type Pokémon are no longer resistant to Dark and Ghost moves.
+#define B_TRAINERS_STEAL_ITEMS      TRUE  // If TRUE, trainer with thief/pickpocket will temporarily steal your items
```
I think this should be reversed, as "Trainers steal items" could mean that they permanetly steal them.

ghoulslash replied:
reversed as in FALSE by default? I can change it to `B_TRAINERS_KNOCK_OFF_ITEMS` too, since that doesn't imply a permanent loss.

AsparagusEduardo replied:
Changing the name of the config should help, but I feel it's not clear enough that the vanilla effect is of permanent loss.
That's why I suggested keeping the current config name but reversing the checks themselves.

AsparagusEduardo commented on src/battle_main.c:
```diff
@@ -4819,6 +4819,10 @@ static void HandleEndTurn_FinishBattle(void)
         sub_8186444();
         BeginFastPaletteFade(3);
         FadeOutMapMusic(5);
+        #if B_TRAINERS_STEAL_ITEMS
```
Same as above

AsparagusEduardo commented on src/battle_script_commands.c:
```diff
@@ -2369,6 +2369,35 @@ static void CheckSetUnburden(u8 battlerId)
+    
+    gBattleStruct->choicedMove[gBattlerTarget] = 0;
+    
+    #if B_TRAINERS_STEAL_ITEMS
```
Same as above

AsparagusEduardo commented on src/battle_script_commands.c:
```diff
@@ -11026,7 +11057,11 @@ static void Cmd_tryswapitems(void) // trick
                                   | BATTLE_TYPE_SECRET_BASE
-                                  | BATTLE_TYPE_x2000000))))
+                                  | BATTLE_TYPE_x2000000
+                                  #if B_TRAINERS_STEAL_ITEMS
```
Same as above

AsparagusEduardo commented on src/battle_script_commands.c:
```diff
@@ -11090,6 +11125,16 @@ static void Cmd_tryswapitems(void) // trick
             PREPARE_ITEM_BUFFER(gBattleTextBuff1, *newItemAtk)
             PREPARE_ITEM_BUFFER(gBattleTextBuff2, oldItemAtk)
+            
+            #if B_TRAINERS_STEAL_ITEMS
```
Same as above.

AsparagusEduardo commented on src/battle_util.c:
```diff
@@ -7764,3 +7764,98 @@ u8 GetBattleMoveSplit(u32 moveId)
+              | BATTLE_TYPE_LINK
+              | BATTLE_TYPE_x2000000
+              | BATTLE_TYPE_SECRET_BASE
+              #if B_TRAINERS_STEAL_ITEMS
```
Same as above

ultima-soul commented on src/battle_util.c:
```diff
@@ -7830,3 +7830,98 @@ u8 GetBattleMoveSplit(u32 moveId)
 }
+
+// sort an array of battlers by speed
+//  useful for effects like pickpocket, eject button, red card, dancer
```
```suggestion
// useful for effects like pickpocket, eject button, red card, dancer
```

ultima-soul commented on src/battle_util.c:
```diff
@@ -7830,3 +7830,98 @@ u8 GetBattleMoveSplit(u32 moveId)
+		battlers[j + 1] = keyBank;
+		speeds[j + 1] = key;
+	}
+}
```
I feel like the terms `key` and `bank` probably need to be renamed to be more clear for what they're supposed to be. Bank especially I believe was an old term used in the IDBs for FR and E which was renamed to `battler` in the Decomps.

ultima-soul reviewed (changes requested):
Here's a gist of my testing notes that suggested things that aren't implemented yet, need to be changed, or are implemented based on USUM testing: https://gist.github.com/ultima-soul/206d137d280ccaa938d709fffe7fa218

ghoulslash commented:
Added sticky barb transfer + interaction with pickpocket:
![pickpocket_stickyBarb](https://user-images.githubusercontent.com/41651341/101664717-6a737400-3a09-11eb-9850-dc91e765779d.gif)

Fixed other concerns, as well

ultima-soul commented on include/battle.h:
```diff
@@ -541,6 +547,7 @@ struct BattleStruct
     u8 sameMoveTurns[MAX_BATTLERS_COUNT]; // For Metronome, number of times the same moves has been SUCCESFULLY used.
     u16 moveEffect2; // For Knock Off
     u16 changedSpecies[PARTY_SIZE]; // For Zygarde or future forms when multiple mons can change into the same pokemon.
+    struct StolenItem itemStolen[PARTY_SIZE];  //player's team that had items stolen (bit per party member)
```
```suggestion
    struct StolenItem itemStolen[PARTY_SIZE];  // Player's team that had items stolen (byte per party member)
```

ultima-soul commented on include/constants/battle_config.h:
```diff
@@ -145,6 +145,7 @@
 #define B_PARALYZE_ELECTRIC         GEN_6 // In Gen6+, Electric-type Pokémon can't be paralyzed.
 #define B_POWDER_GRASS              GEN_6 // In Gen6+, Grass-type Pokémon are immune to powder and spore moves.
 #define B_STEEL_RESISTANCES         GEN_6 // In Gen6+, Steel-type Pokémon are no longer resistant to Dark and Ghost moves.
+#define B_TRAINERS_KNOCK_OFF_ITEMS  TRUE  // If TRUE, trainers can steal/swap your items (non-berries are restored after battle). In vanilla games trainers cannot steal items.
```
Comment is incorrect as trainers can steal items in vanilla, just not permanently unless the item was consumed by them. I noted that the setting being TRUE matches the vanilla behaviour when testing for my previous review.

ultima-soul commented on src/battle_util.c:
```diff
@@ -5862,6 +5850,22 @@ u8 ItemBattleEffects(u8 caseID, u8 battlerId, bool8 moveTurn)
+                  && IsBattlerAlive(gBattlerAttacker)
+                  && gBattleMons[gBattlerAttacker].item == ITEM_NONE)
+                {
+                    //no sticky hold checks. item is already known so no CanStealItem checks
+                    gEffectBattler = battlerId; //effect battler = target
+                    StealTargetItem(gBattlerAttacker, gBattlerTarget);  //attacker takes target's barb
```
```suggestion
                    // No sticky hold checks. item is already known so no CanStealItem checks
                    gEffectBattler = battlerId; // gEffectBattler = target
                    StealTargetItem(gBattlerAttacker, gBattlerTarget);  // Attacker takes target's barb
```

ultima-soul commented on src/battle_util.c:
```diff
@@ -5890,6 +5894,18 @@ u8 ItemBattleEffects(u8 caseID, u8 battlerId, bool8 moveTurn)
                 RecordItemEffectBattle(battlerId, battlerHoldEffect);
             }
             break;
+        case HOLD_EFFECT_STICKY_BARB:   //not an orb per-say, but similar effect, and needs to NOT activate with pickpocket
```
```suggestion
        case HOLD_EFFECT_STICKY_BARB:   // Not an orb per se, but similar effect, and needs to NOT activate with pickpocket
```

ultima-soul commented on src/battle_util.c:
```diff
@@ -7830,3 +7858,118 @@ u8 GetBattleMoveSplit(u32 moveId)
 }
+
+// sort an array of battlers by speed
+//  useful for effects like pickpocket, eject button, red card, dancer
```
```suggestion
// useful for effects like pickpocket, eject button, red card, dancer
```

ultima-soul commented on src/battle_util.c:
```diff
@@ -7830,3 +7858,118 @@ u8 GetBattleMoveSplit(u32 moveId)
+        {
+            stolenItem = gBattleStruct->itemStolen[i].originalItem;
+            if (stolenItem != ITEM_NONE && ItemId_GetPocket(stolenItem) != POCKET_BERRIES)
+                SetMonData(&gPlayerParty[i], MON_DATA_HELD_ITEM, &stolenItem);  //restore stolen non-berry items
```
```suggestion
                SetMonData(&gPlayerParty[i], MON_DATA_HELD_ITEM, &stolenItem);  // Restore stolen non-berry items
```

ultima-soul reviewed (changes requested):
I noted the following things that need to be implemented:
```
- Pickpocket can't steal Griseous Orb if attacker Pokemon is Giratina Origin Forme
- Pickpocket can't steal Plate if attacker Pokemon is Arceus non-Normal Forme
- Pickpocket can't steal Memory if attacker Pokemon is Silvally non-Normal Forme
- Pickpocket can't steal Drive if attacker Pokemon is Genesect non-base Form
- Pickpocket can't steal Red Orb if Pickpocketer Pokemon is Groudon
- Pickpocket can't steal Red Orb if Pickpocketer Pokemon is Primal Groudon
- Pickpocket can't steal Blue Orb if Pickpocketer Pokemon is Kyogre
- Pickpocket can't steal Blue Orb if Pickpocketer Pokemon is Primal Kyogre
- Pickpocket can't steal Griseous Orb if Pickpocketer Pokemon is Giratina
- Pickpocket can't steal Griseous Orb if Pickpocketer Pokemon is Giratina Origin Forme
- Pickpocket can't steal Plate if Pickpocketer Pokemon is Arceus
- Pickpocket can't steal Plate if Pickpocketer Pokemon is Arceus non-Normal Forme
- Pickpocket can't steal Memory if Pickpocketer Pokemon is Silvally
- Pickpocket can't steal Memory if Pickpocketer Pokemon is Silvally non-Normal Forme
- Pickpocket can't steal Drive if Pickpocketer Pokemon is Genesect
- Pickpocket can't steal Drive if Pickpocketer Pokemon is Genesect non-base form
- Pickpocket can't steal Mega Stone if Pickpocketer Pokemon is Mega Evolvable with that Mega Stone
- Pickpocket can't steal Mega Stone if Pickpocketer Pokemon is Mega Evolution that could have evolved with that Mega Stone (Charizard X/Y and Mewtwo X/Y also fail to steal if it's the other Mega Stone its base form could've used)
- Items stolen from trainers are not kept when battle ends.
```

ghoulslash commented:
Ready for re-review.

TESTED:
- stealing leftovers from trainer -> kept after battle
- stealing non-compatible mega stone (thief & pickpocket) -> works, keeps after trainer battle
- stealing compatible mega stone (thief & pickpocket) -> does not work
- by extension, all other specific checks will work/fail appropriately based on `CanBattlerGetOrLoseItem`
- sticky barb with FALSE config -> does not transfer to trainer, kept after battle
- sticky barb with TRUE config -> transfers, (pickpocket interaction still good), sticky barb returned after battle

ghoulslash commented:
added base form checks to `CanBattlerGetOrLoseItem`, ready for review

ultima-soul commented on src/battle_util.c:
```diff
@@ -7835,25 +7852,32 @@ bool32 DoBattlersShareType(u32 battler1, u32 battler2)
         return FALSE;
-    else if (species == SPECIES_KYOGRE && itemId == ITEM_BLUE_ORB)
+    else if (GET_BASE_SPECIES_ID(species) == SPECIES_KYOGRE && (itemId == ITEM_BLUE_ORB || holdEffect == HOLD_EFFECT_PRIMAL_ORB))
         return FALSE;
-    else if (species == SPECIES_GROUDON && itemId == ITEM_RED_ORB)
+    else if (GET_BASE_SPECIES_ID(species) == SPECIES_GROUDON && (itemId == ITEM_BLUE_ORB || holdEffect == HOLD_EFFECT_PRIMAL_ORB))
```
Groudon's check should use ITEM_RED_ORB.

Also, isn't it a bad idea to do an OR check with HOLD_EFFECT_PRIMAL_ORB because then it would make it so that Kyogre can't steal a Red Orb or lose it and vice versa for Groudon?

ghoulslash replied:
fixed the Red Orb issue. Primal orbs use are red/blue orb so I am assuming for now that just checking these is sufficient

ultima-soul commented on src/battle_util.c:
```diff
@@ -7835,25 +7852,32 @@ bool32 DoBattlersShareType(u32 battler1, u32 battler2)
-    else if (species == SPECIES_GROUDON && itemId == ITEM_RED_ORB)
+    else if (GET_BASE_SPECIES_ID(species) == SPECIES_GROUDON && itemId == ITEM_RED_ORB) // includes primal
         return FALSE;
     // Mega stone cannot be lost if pokemon can mega evolve with it or is already mega evolved.
```
```suggestion
    // Mega stone cannot be lost if pokemon's base species can mega evolve with it.
```

ultima-soul commented on src/battle_util.c:
```diff
@@ -7835,25 +7852,32 @@ bool32 DoBattlersShareType(u32 battler1, u32 battler2)
     // Mega stone cannot be lost if pokemon can mega evolve with it or is already mega evolved.
-    else if (ItemId_GetHoldEffect(itemId) == HOLD_EFFECT_MEGA_STONE
-             && ((GetMegaEvolutionSpecies(species, itemId) != SPECIES_NONE) || gBattleStruct->mega.evolvedPartyIds[GetBattlerSide(battlerId)] & gBitTable[gBattlerPartyIndexes[battlerId]]))
+    else if (holdEffect == HOLD_EFFECT_MEGA_STONE
+      && ((GetMegaEvolutionSpecies(species, itemId) != SPECIES_NONE)
+      || gBattleStruct->mega.evolvedPartyIds[GetBattlerSide(battlerId)] & gBitTable[gBattlerPartyIndexes[battlerId]]))
```
```suggestion
    else if (GetMegaEvolutionSpecies(GET_BASE_SPECIES_ID(species), itemId) != SPECIES_NONE 
             && holdEffect == HOLD_EFFECT_MEGA_STONE)
```
to fix not being implemented:
```
- Pickpocket can't steal Mega Stone if attacker Pokemon is Mega Evolution of species that could have evolved with that Mega Stone
- Pickpocket can't steal Mega Stone if Pickpocketer Pokemon is Mega Evolution of species that could have evolved with that Mega Stone
```
For example, Mega Charizard X that is holding Charizardite X or Y cannot have it taken, but if holding Diancite can have it stolen from it. Another example: Mega Charizard X with Pickpocket cannot steal Charizardite X or Y from contact attacker, but can steal Diancite.

ultima-soul commented on src/battle_util.c:
```diff
@@ -7835,25 +7852,32 @@ bool32 DoBattlersShareType(u32 battler1, u32 battler2)
+    else if (GET_BASE_SPECIES_ID(species) == SPECIES_GIRATINA && itemId == ITEM_GRISEOUS_ORB)
         return FALSE;
-    else if (species == SPECIES_GENESECT && GetBattlerHoldEffect(battlerId, FALSE) == HOLD_EFFECT_DRIVE)
+    else if (GET_BASE_SPECIES_ID(species) == SPECIES_GENESECT && GetBattlerHoldEffect(battlerId, FALSE) == HOLD_EFFECT_DRIVE)
```
```suggestion
    else if (GET_BASE_SPECIES_ID(species) == SPECIES_GENESECT && holdEffect == HOLD_EFFECT_DRIVE)
```
to fix not being implemented:
```
- Pickpocket can't steal Drive if Pickpocketer Pokemon is Genesect (All Forms)
```

ultima-soul commented on src/battle_util.c:
```diff
@@ -7835,25 +7852,32 @@ bool32 DoBattlersShareType(u32 battler1, u32 battler2)
+    else if (GET_BASE_SPECIES_ID(species) == SPECIES_GENESECT && GetBattlerHoldEffect(battlerId, FALSE) == HOLD_EFFECT_DRIVE)
         return FALSE;
-    else if (species == SPECIES_SILVALLY && GetBattlerHoldEffect(battlerId, FALSE) == HOLD_EFFECT_MEMORY)
+    else if (GET_BASE_SPECIES_ID(species) == SPECIES_SILVALLY && GetBattlerHoldEffect(battlerId, FALSE) == HOLD_EFFECT_MEMORY)
```
```suggestion
    else if (GET_BASE_SPECIES_ID(species) == SPECIES_SILVALLY && holdEffect == HOLD_EFFECT_MEMORY)
```
to fix not being implemented:
```
- Pickpocket can't steal Memory if Pickpocketer Pokemon is Silvally (All Forms)
```

ultima-soul commented on src/battle_util.c:
```diff
@@ -7835,25 +7852,32 @@ bool32 DoBattlersShareType(u32 battler1, u32 battler2)
-        return FALSE;
-    else if (itemId == ITEM_ENIGMA_BERRY)
+    u16 holdEffect = ItemId_GetHoldEffect(itemId);
+    
+    // Mail can be stolen now
```
```suggestion
```
Kinda unnecessary comment since it doesn't pertain to actual code anymore.

ultima-soul commented on include/constants/battle_config.h:
```diff
@@ -144,6 +144,9 @@
 #define B_CATCHING_CHARM_BOOST      20    // % boost in Critical Capture odds if player has the Catching Charm.
 
+// Item Theft Settings
+#define B_TRAINERS_KNOCK_OFF_ITEMS  TRUE  // If TRUE, trainers can steal/swap your items (non-berries are restored after battle). In vanilla games trainers cannot steal items.
```
This should maybe be turned into a Generational config because Gen 2-4 doesn't allow wild mons or regular in-game trainers to steal items apparently and Gen 5+ allows it. The config name should also be changed to something more like `B_TRAINERS_STEAL_ITEMS`.

ultima-soul commented on include/battle.h:
```diff
@@ -542,6 +548,7 @@ struct BattleStruct
     u8 sameMoveTurns[MAX_BATTLERS_COUNT]; // For Metronome, number of times the same moves has been SUCCESFULLY used.
     u16 moveEffect2; // For Knock Off
     u16 changedSpecies[PARTY_SIZE]; // For Zygarde or future forms when multiple mons can change into the same pokemon.
+    struct StolenItem itemStolen[PARTY_SIZE];  // Player's team that had items stolen (byte per party member)
```
```suggestion
    struct StolenItem itemStolen[PARTY_SIZE];  // Player's team that had items stolen (two bytes per party member)
```

ultima-soul reviewed (changes requested):
A couple of other issues:
- Pickpocket doesn't activate after user uses up its Focus Sash and has Max 1 HP
It does activate if the Pickpocketer with Focus Sash has more than 1 Max HP so this is a strange case.

- Pickpocket doesn't activate after Knock Off makes Pickpocketer lose item
This might be related to Knock Off being the old effect which is just disabling the item? Fwiw, Bug Bite and Pluck work as expected.

- Items stolen from wild Pokemon are kept as expected, but Gen 5+ prevents you from permanently stealing from NPC trainers which isn't the case right now. Right now you keep the item you steal from a trainer.

ghoulslash commented:
> * Pickpocket doesn't activate after user uses up its Focus Sash and has Max 1 HP
>   It does activate if the Pickpocketer with Focus Sash has more than 1 Max HP so this is a strange case.

This sounds like focus sash and shedinja problem, not a pickpocket problem.

> 
> * Pickpocket doesn't activate after Knock Off makes Pickpocketer lose item
>   This might be related to Knock Off being the old effect which is just disabling the item? Fwiw, Bug Bite and Pluck work as expected.

This is an issue with how Knock off is handled in Gen3,4. This then causes conflict with which item the pickpocketer would keep after battle. The knocked off one or the one it stole? Either way, it feels out of scope with this PR. That should be a separate knock off update.

> * Items stolen from wild Pokemon are kept as expected, but Gen 5+ prevents you from permanently stealing from NPC trainers which isn't the case right now. Right now you keep the item you steal from a trainer.

I suppose I can make this conditional.
